<robot xmlns:xacro="http://ros.org/wiki/xacro"
       xmlns:sensor="http://playerstage.sourceforge.net/gazebo/xmlschema/#sensor"
       xmlns:controller="http://playerstage.sourceforge.net/gazebo/xmlschema/#controller"
       xmlns:joint="http://playerstage.sourceforge.net/gazebo/xmlschema/#slider"
       xmlns:interface="http://playerstage.sourceforge.net/gazebo/xmlschema/#interface"
       xmlns:body="http://playerstage.sourceforge.net/gazebo/xmlschema/#body"
       xmlns:geom="http://playerstage.sourceforge.net/gazebo/xmlschema/#geom">

  <xacro:include filename="$(find sr_description)/hand/xacro/thumb/thdistal.gazebo.xacro" />
  <xacro:include filename="$(find sr_description)/hand/xacro/thumb/thdistal.transmission.xacro" />
  <!-- muscle is true for muscle hand or false for motor hand -->
	<!-- bio is true for biotac finger or false for standard finger -->
  <!-- ubi is true for ubi finger or false for standard finger -->
  <!-- eli is true for ellipsoid fingertip or false for standard finger -->
  <!-- prefix can be anything but usually is "r_" or  "l_" for right and left hands distinction -->
  <!-- parent is the parent link for the first joint -->
  <xacro:macro name="thdistal" params="muscle bio ubi eli prefix parent">  
    
		
		<xacro:if value="${bio}">
			<!--	BIOTAC -->
			<link name="${prefix}thbiotac_J1_dummy">
				<inertial>
					<origin xyz="0 0 0" rpy="0 0 0" />
					<mass value="0.01" />
					 <inertia  ixx="0.0007" ixy="0.0"  ixz="0.0"  iyy="0.0007"  iyz="0.0"
        izz="0.00010" />
        </inertial>
			</link>
			
			<joint name="${prefix}THJ1" type="revolute">
				<parent link="${parent}"/>
				<child link="${prefix}thbiotac_J1_dummy"/>
				<origin xyz="0 0 0.02" rpy="0 0 0" />
				<axis xyz="1 0 0" />
				<limit lower="${20*M_PI/180}" upper="${20.1*M_PI/180}" effort="10" velocity="0.1" />
				<dynamics damping="0.001"/> 
			</joint>
			
			<link name="${prefix}thbiotac">
				<inertial>
					<origin xyz="0 0 0" rpy="0 0 0" />
					<mass value="0.2" />
					 <inertia  ixx="0.0007" ixy="0.0"  ixz="0.0"  iyy="0.0007"  iyz="0.0"
        izz="0.00010" />
				</inertial>
				<visual>
					<origin xyz="0 0 0" rpy="0 0 0" />
					<geometry name="${prefix}th_biotac_visual">
						<!-- <mesh filename="package://sr_description/hand/model/biotacs/biotac.stl" scale="0.0254 0.0254 0.0254" /> -->
						<mesh filename="package://sr_description/hand/model/biotacs/biotac_decimated.dae" scale="0.0254 0.0254 0.0254" />
					</geometry>
					<material name="BiotacGreen" />
				</visual>
				<collision>
					<geometry name="${prefix}th_biotac_collision_geom">
						<!-- <box size="0.001 0.001 0.001" /> -->
						<!-- <mesh filename="package://sr_description/hand/model/biotacs/biotac-v5.stl" scale="0.0254 0.0254 0.0254" /> -->
						<mesh filename="package://sr_description/hand/model/biotacs/biotac_decimated.dae" scale="0.0254 0.0254 0.0254" />
					</geometry>
				</collision>
			</link>

			<gazebo reference="${prefix}thdistal">
				<sensor type="contact" name="${prefix}thdistal_contact">
					<always_on>1</always_on>
					<update_rate>1000.0</update_rate>
					<contact>
						<collision>${prefix}thbiotac_geom</collision>
						<topic>${prefix}thbiotac_collision</topic>
					</contact>
					<plugin name="${prefix}thbiotac_gazebo_ros_bumper_controller" filename="libgazebo_ros_bumper.so">
						<alwaysOn>true</alwaysOn>
						<frameName>${prefix}thbiotac</frameName>
						<updateRate>100.0</updateRate>
						<bumperTopicName>contacts/${prefix}th/distal</bumperTopicName>
					</plugin>
				</sensor>
				<selfCollide>true</selfCollide>
				<material>BiotacGreen</material>
			</gazebo>

			<joint name="${prefix}th_biotac_joint" type="fixed">
				<parent link="${prefix}thmiddle"/>
				<child link="${prefix}thbiotac"/>
				<origin xyz="0 0.0 0.01" rpy="0 0 -1.571" />
			</joint>
				
			
		  <!-- extra link to imaginary sphere in fingertip for FK/IK calculations -->
			<!-- remove visual and collision once the correct position is found -->
			<link name="${prefix}thtip">
				<inertial>
				 <mass value="0.001" />
				 <origin xyz="0 0 0" rpy="0 0 0" />
				 <inertia  ixx="0" ixy="0"  ixz="0"  iyy="0"  iyz="0" izz="0" />
				</inertial>
				<visual>
					<geometry>
						<sphere radius="0.002" />
					</geometry>
				</visual>
				<collision>
					<geometry>
						<box size="0.001 0.001 0.001" />
					</geometry>
				</collision>
			</link>

			<gazebo reference="${prefix}thtip">
				<material name="LightGrey"/>
			</gazebo>

			<joint name="${prefix}thtip" type="fixed">
				<parent link="${prefix}thmiddle"/>
				<child link="${prefix}thtip"/>
				<origin xyz="-0.009 0.0 ${0.01+0.044}" rpy="0 0 ${-M_PI/2}" /> <!-- approximated only FIXME -->
			</joint>	
		</xacro:if>
		
			<xacro:unless value="${bio}">
				<!-- common -->
				<link name="${prefix}thdistal">
				<inertial>
					<mass value="0.016" />
					<origin xyz="0 0 0.01375" rpy="0 0 0"/>
					 <inertia  ixx="0.0007" ixy="0.0"  ixz="0.0"  iyy="0.0007"  iyz="0.0"
        izz="0.00010" />
				</inertial>
				<visual>
					<origin xyz="0 0 0" rpy="0 0 0" />
					<geometry name="${prefix}thdistal_visual">
						<xacro:if value="${eli}">
							<!-- Ellispoid distal  -->
							<mesh filename="package://sr_description/hand/model/ellipsoids/thdistal_ellipsoid_visual.dae" scale="0.001 0.001 0.001" />
						</xacro:if> 
						<xacro:unless value="${eli}">
							<xacro:if value="${ubi}">
								<!-- Ubi distal -->
								<mesh filename="package://sr_description/hand/model/TH1_z.dae" scale="0.001 0.001 0.001" />
							</xacro:if> 
							<!-- Standard distal  -->
							<xacro:unless value="${ubi}">
								<mesh filename="package://sr_description/hand/model/TH1_z.dae" scale="0.001 0.001 0.001" />
							</xacro:unless> 
						</xacro:unless> 
					</geometry>
					<material name="shadow_thmiddle_material">
						<color rgba="0.2 0.2 0.2 1.0"/>
					</material>
				</visual>
				<collision>
					<origin xyz="0 0 0" rpy="0 0 0 " />
					<geometry name="${prefix}thmiddle_collision_geom">
						<xacro:if value="${eli}">
							<!-- Ellispoid distal  -->
						  <mesh filename="package://sr_description/hand/model/ellipsoids/thdistal_ellipsoid.dae" scale="0.001 0.001 0.001" />
						</xacro:if> 
						<xacro:unless value="${eli}">
							<xacro:if value="${ubi}">
								<!-- Ubi distal -->
								<mesh filename="package://sr_description/hand/model/TH1_z.dae" scale="0.001 0.001 0.001" />
							</xacro:if> 
							<xacro:unless value="${ubi}">
								<!-- Standard distal  -->
								<mesh filename="package://sr_description/hand/model/TH1_z.dae" scale="0.001 0.001 0.001" />
							</xacro:unless> 
						</xacro:unless> 						
						<!--	  <box size="0.015 0.015 0.015" /> -->
					</geometry>
				</collision>
			</link>
    
			
			<joint name="${prefix}THJ1" type="revolute">
				<parent link="${parent}"/>
				<child link="${prefix}thdistal"/>
				<origin xyz="0 0 0.032" rpy="0 0 ${-M_PI/2}" />
				<axis xyz="1 0 0" />
				<limit lower="0" upper="${90.0*M_PI/180}" effort="10" velocity="1.0" />
				<dynamics damping="0.001"/> 
			</joint>
			<link name="${prefix}thtip">
				<inertial>
					<mass value="0.001" />
					<inertia  ixx="0.0" ixy="0.0"  ixz="0.0"  iyy="0.0"  iyz="0.0" izz="0.0" />
				</inertial>
			</link>
			<joint name="${prefix}thtip" type="fixed">
				<parent link="${prefix}thdistal"/>
				<child link="${prefix}thtip"/>
				<origin xyz="0 0 0.0275" rpy="0 0 0" />
			</joint>
			
			<xacro:thdistal_gazebo prefix="${prefix}" />
		</xacro:unless> 
  
		<!-- extensions -->
		<xacro:thdistal_transmission muscletrans="${muscle}" prefix="${prefix}"/>

  </xacro:macro>

</robot>
